sdk/storage/azure-storage-common/src/crypt.cpp

```diff
@@ -15,91 +15,59 @@
 #include <openssl/sha.h>
 #endif
 
+#include <algorithm>
 #include <stdexcept>
 #include <vector>
 
+#include "azure/core/http/http.hpp"
 #include "azure/storage/common/storage_common.hpp"
 
 namespace Azure { namespace Storage {
 
   namespace Details {
-    static const char* c_unreserved
-        = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-._~";
     static const char* c_subdelimiters = "!$&'()*+,;=";
-    const char* c_hex = "0123456789ABCDEF";
 
     std::string UrlEncodeQueryParameter(const std::string& value)
     {
-      const static std::vector<bool> shouldEncodeTable = []() {
-        std::string queryCharacters
-            = std::string(c_unreserved) + std::string(c_subdelimiters) + "%/:@?";
-
-        std::vector<bool> ret(256, true);
-        for (char c : queryCharacters)
-        {
-          ret[c] = false;
-        }
-        // we also encode % and +
-        ret['%'] = true;
-        ret['+'] = true;
-        // Surprisingly, '=' also needs to be encoded because Azure Storage server side is so
-        // strict. We are applying this function to query key and value respectively, so this won't
-        // affect that = used to separate key and query.
-        ret['='] = true;
-        return ret;
+      const static std::string c_doNotEncodeCharacters = []() {
+        // Core::Http::Url::Encode won't encode unreserved characters.
+        std::string doNotEncodeCharacters = c_subdelimiters;
+        doNotEncodeCharacters += "%/:@?";
+        doNotEncodeCharacters.erase(
+            std::remove_if(
+                doNotEncodeCharacters.begin(),
+                doNotEncodeCharacters.end(),
+                [](char x) {
+                  // we also encode % and +
+                  // Surprisingly, '=' also needs to be encoded because Azure Storage server side is
+                  // so strict. We are applying this function to query key and value respectively,
+                  // so this won't affect that = used to separate key and query.
+                  return x == '%' || x == '+' || x == '=';
+                }),
+            doNotEncodeCharacters.end());
+        return doNotEncodeCharacters;
```

Comment:
This is a lot of logical complexity. Leverage the Encode API to simplify this like https://github.com/Azure/azure-sdk-for-cpp/issues/736#issuecomment-710733090

```C++
// Most of the sub-delimiters don't need to be encoded. 
// However, '=' also needs to be encoded because Azure Storage server side is so strict. 
// We are applying this function to query key and value respectively, 
// so this won't affect that = used to separate key and query.
string doNotEncodeQueryParameter = "!$&'()*,;/:@?";
return Core::Http::Url::Encode(value, doNotEncodeQueryParameter);
```

Reply:
people may wonder where this constant string `!$&'()*,;/:@?` comes.

Reply:
The same could be said about things like ` "/:@?"` that you have currently. A comment might be a better way to solve that problem (the subdelimiters need to be encoded other than these exceptions: ...).

Reply:
I still think subdelimiters should be defined separately as a const string, as it is clearly defined in RFC.

Reply:
Instead of `doNotEncodeQueryParameter`, it should be called `storageShouldNotHttpEncodeCharactersInQueryParameter` to be self-explanatory. I don't think the former name would clarify much for the customer and the latter name looks ridiculous. With this said, I do agree that we should have separate const defined for RFC regulated delimiter and Storage specific exceptions.

Reply:
To re-iterate, this implementation is unnecessarily complicated and in an attempt to be "self-documenting" it is less maintainable. The use of the `c_subdelimiters` make it  seem like this is more general purpose as a Url encoder, which doesn't fit the domain this is being used/called in, it is specific to just storage service APIs. There isn't a clean separation of duties here.